tools/circt-test/circt-test.cpp

```diff
@@ -50,58 +50,122 @@ namespace {
 /// The tool's command line options.
 struct Options {
   cl::OptionCategory cat{"circt-test Options"};
+
   cl::opt<std::string> inputFilename{cl::Positional, cl::desc("<input file>"),
                                      cl::init("-"), cl::cat(cat)};
+
   cl::opt<std::string> outputFilename{
       "o", cl::desc("Output filename (`-` for stdout)"),
       cl::value_desc("filename"), cl::init("-"), cl::cat(cat)};
+
+  cl::opt<bool> listTests{"l", cl::desc("List tests in the input and exit"),
+                          cl::init(false), cl::cat(cat)};
+
   cl::opt<bool> json{"json", cl::desc("Emit test list as JSON array"),
                      cl::init(false), cl::cat(cat)};
 };
 Options opts;
 
 } // namespace
 
+//===----------------------------------------------------------------------===//
+// Test Discovery
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// The various kinds of test that can be executed.
+enum class TestKind { Formal };
+
+/// A single discovered test.
+class Test {
+public:
+  /// The name of the test. This is also the name of the top-level module passed
+  /// to the formal or simulation tool to be run.
+  StringAttr name;
+  /// The kind of test, such as "formal" or "simulation".
+  TestKind kind;
+  /// An optional location indicating where this test was discovered. This can
+  /// be the location of an MLIR op, or a line in some other source file.
+  LocationAttr loc;
```

Comment:
nit: `op.getLoc()` return `Location` which a non-null wrapper around LocationAttr so maybe just use `Location` here.

Reply:
Good point. I think there will be cases in the future where we don't have a location. For example, if we discover tests by looking for something like "// RUN: ..." lines in SystemVerilog files, we may not have a reliable location. Or maybe we do?